File: CS/WindowsApplication1/CustomEditor/RepositoryItemMyLookUpEdit.cs
using System;
using System.ComponentModel;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using System.Collections.Generic;

namespace WindowsApplication1 {
    public delegate void GetAutoCompleteListEventHandler(object sender, AutoCompleteListEventArgs e);

    [System.ComponentModel.DesignerCategory("")]
    [UserRepositoryItem("Register")]
    public class RepositoryItemMyLookUpEdit : RepositoryItemLookUpEdit {
        private static readonly object getAutoCompleteList = new object();

        public event GetAutoCompleteListEventHandler GetAutoCompleteList {
            add {
                Events.AddHandler(getAutoCompleteList, value);
                OnPropertiesChanged();
            }
            remove {
                Events.RemoveHandler(getAutoCompleteList, value);
                OnPropertiesChanged();
            }
        }


        /// <summary>
        /// Triggers the GetAutoCompleteList event.
        /// </summary>
        public virtual void RaiseGetAutoCompleteList(AutoCompleteListEventArgs ea) {
            GetAutoCompleteListEventHandler handler = (GetAutoCompleteListEventHandler)Events[getAutoCompleteList];
            if (handler != null) {
                handler(GetEventSender(), ea);
            }
        }


        static RepositoryItemMyLookUpEdit() {
            Register();
        }
        public RepositoryItemMyLookUpEdit() {
            DataSource = new List<string>();
            ImmediatePopup = true;
            TextEditStyle = DevExpress.XtraEditors.Controls.TextEditStyles.Standard;
        }
        internal const string EditorName = "MyLookUpEdit";

        public static void Register() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(MyLookUpEdit),
                typeof(RepositoryItemMyLookUpEdit), typeof(DevExpress.XtraEditors.ViewInfo.LookUpEditViewInfo),
                new DevExpress.XtraEditors.Drawing.ButtonEditPainter(), true, null));
        }
        public override string EditorTypeName {
            get {
                return EditorName;
            }
        }
        public void UpdateDataSource(string autoSearchText) {
            List<string> items = DataAdapter.DataSource as List<string>;
            RaiseGetAutoCompleteList(new AutoCompleteListEventArgs(autoSearchText, items));
        }
        public override void Assign(RepositoryItem item) {
            BeginUpdate();
            try {
                base.Assign(item);
                RepositoryItemMyLookUpEdit source = item as RepositoryItemMyLookUpEdit;
                if(source == null)
                    return;
                Events.AddHandler(getAutoCompleteList, source.Events[getAutoCompleteList]);
            } finally {
                EndUpdate();
            }
        }
    }
}

File: VB/WindowsApplication1/CustomEditor/RepositoryItemMyLookUpEdit.vb
Imports System
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports System.Collections.Generic

Namespace WindowsApplication1
	Public Delegate Sub GetAutoCompleteListEventHandler(ByVal sender As Object, ByVal e As AutoCompleteListEventArgs)

	<System.ComponentModel.DesignerCategory(""), UserRepositoryItem("Register")>
	Public Class RepositoryItemMyLookUpEdit
		Inherits RepositoryItemLookUpEdit

		Private Shared ReadOnly getAutoCompleteListObj As New Object()

		Public Custom Event GetAutoCompleteList As GetAutoCompleteListEventHandler
			AddHandler(ByVal value As GetAutoCompleteListEventHandler)
				Events.AddHandler(getAutoCompleteListObj, value)
				OnPropertiesChanged()
			End AddHandler
			RemoveHandler(ByVal value As GetAutoCompleteListEventHandler)
				Events.RemoveHandler(getAutoCompleteListObj, value)
				OnPropertiesChanged()
			End RemoveHandler
			RaiseEvent(ByVal sender As Object, ByVal e As AutoCompleteListEventArgs)
			End RaiseEvent
		End Event


		''' <summary>
		''' Triggers the GetAutoCompleteList event.
		''' </summary>
		Public Overridable Sub RaiseGetAutoCompleteList(ByVal ea As AutoCompleteListEventArgs)
			Dim handler As GetAutoCompleteListEventHandler = CType(Events(getAutoCompleteListObj), GetAutoCompleteListEventHandler)
			If handler IsNot Nothing Then
				handler(GetEventSender(), ea)
			End If
		End Sub


		Shared Sub New()
			Register()
		End Sub
		Public Sub New()
			DataSource = New List(Of String)()
			ImmediatePopup = True
			TextEditStyle = DevExpress.XtraEditors.Controls.TextEditStyles.Standard
		End Sub
		Friend Const EditorName As String = "MyLookUpEdit"

        Public Shared Sub Register()
            Dim img As Image = Nothing
            Dim editorClassInfo As Object = New EditorClassInfo(EditorName, GetType(MyLookUpEdit), GetType(RepositoryItemMyLookUpEdit), GetType(DevExpress.XtraEditors.ViewInfo.LookUpEditViewInfo), New DevExpress.XtraEditors.Drawing.ButtonEditPainter(), True, img)
            EditorRegistrationInfo.Default.Editors.Add(editorClassInfo)
        End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return EditorName
			End Get
		End Property
		Public Sub UpdateDataSource(ByVal autoSearchText As String)
			Dim items As List(Of String) = TryCast(DataAdapter.DataSource, List(Of String))
			RaiseGetAutoCompleteList(New AutoCompleteListEventArgs(autoSearchText, items))
		End Sub
		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As RepositoryItemMyLookUpEdit = TryCast(item, RepositoryItemMyLookUpEdit)
				If source Is Nothing Then
					Return
				End If
				Events.AddHandler(getAutoCompleteListObj, source.Events(getAutoCompleteListObj))
			Finally
				EndUpdate()
			End Try
		End Sub
	End Class
End Namespace
